### ISSUE TITLE
When should cryptolib disable icache to keep operations constant-time?

### ISSUE THREAD

**Issue by drewmacrae** [lowRISC/opentitan#15832]
Some constant-time cryptolib tests are failing when the icache is enabled by the test-rom. 

Does the threat model include the timing of cryptolib code (in which case it should disable the icache when running)
or do we expect that previous accesses to cryptolib were made by similarly trusted code and we should prioritize the performance improvements offerred by the icache?

**Comment by jadephilipoom** [comment]
Yes, the threat model includes timing attacks. As discussed in #15788, we should disable icache upon entry to all constant-time cryptolib functions, and re-enable it before returning. Thanks for raising this issue @drewmacrae !

**Comment by tjaychen** [comment]
@jadephilipoom just a quick question, does the icache timing usually end up being a big threat vector? I definitely understand it can be for data cache, but just wondering if the impact is less pronounced with icache, or is it every bit as dangerous? 

@drewmacrae if we do plan on enabling / disabling icache dynamically, we probably also want to make sure we have test coverage for this somewhere just to check the system is stable when doing so.

**Comment by jadephilipoom** [comment]
> @jadephilipoom just a quick question, does the icache timing usually end up being a big threat vector? I definitely understand it can be for data cache, but just wondering if the impact is less pronounced with icache, or is it every bit as dangerous?

I think at any rate we should be very careful, and the fact that a constant-time test started failing because the timing was variable with icache on (#15788) is a worrying sign. I think to enable it we need some thorough evaluation of how to use it safely for crypto code and also how to evaluate that it's safe (e.g. if it just produces some harmless noise in cycle-counts, how do we check that things are constant-time in tests?).

**Comment by tjaychen** [comment]
i see what you mean. So in the absence of a good way to really evaluate we are "safe", it's better to just be on the conservative side and make sure constant time really means constant time.

**Comment by cdgori** [comment]
In theory, we might be able to flush/invalidate the whole icache before running cryptolib functions.  We should get deterministic cycle counts then.  (Some questions arise about interrupts / context switches though.)

Erring on the side of conservative here is a good idea.

**Comment by msfschaffner** [comment]
@jadephilipoom are any further HW knobs required to support this, or do we have everything in place today?

**Comment by jadephilipoom** [comment]
Well, I could actually use some help from the HW side here. I've tried to call `icache_invalidate()` to insert `fence.i` instructions for the AES-GCM timing tests mentioned in #15788, but so far I still don't get consistent timing. I don't know if this is because of a genuine error with code being not constant-time, or because the icache isn't getting fully invalidated, or if simply invalidating the icache is not enough and we need a stronger "disable icache" knob.

**Comment by msfschaffner** [comment]
@GregAC any ideas?

**Comment by johannheyszl** [comment]
#27678 adds a function to disable/enable iCache and adds it upon executing AES-GCM.

**Comment by johannheyszl** [comment]
Closing this since it seems disabling for AES-GCM is sufficient.
